### PR TITLE
[BEAM-7093] Support Spark 3 in Spark runner

### PR DESCRIPTION
**Pull request by iemejia**
For early review
R: @ibzib

### CODE REVIEW

File: sdks/java/core/build.gradle [hunk 1]
```diff
@@ -69,6 +69,13 @@ test {
   }
 }
 
+// Avoid paranamer 2.7 from leaking into the user classpath. Paranamer is an
+// Avro transitive dependency that can produce runtime conflicts between
+// versions of Spark/Avro
+configurations.shadow {
```
Comment (ibzib) [inline sdks/java/core/build.gradle:75]: I think there's a better way to do this, by changing avro something like this in the `dependencies` block:

```
  shadow(library.java.avro) {
    exclude group: ....
  }
```

That makes clearer where the dependency is coming from. Though I guess it wouldn't prevent paranamer from coming in through other dependencies.

File: .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Spark.groovy [hunk 2]
```diff
@@ -40,10 +40,16 @@ PostcommitJobBuilder.postCommitJob('beam_PostCommit_XVR_Spark',
           shell("echo \"*** RUN CROSS-LANGUAGE SPARK USING PYTHON ${pythonVersion} ***\"")
           gradle {
             rootBuildScriptDir(commonJobProperties.checkoutDir)
-            tasks(':runners:spark:job-server:validatesCrossLanguageRunner')
+            tasks(':runners:spark:2:job-server:validatesCrossLanguageRunner')
             commonJobProperties.setGradleSwitches(delegate)
             switches("-PpythonVersion=${pythonVersion}")
           }
+          gradle {
+              rootBuildScriptDir(commonJobProperties.checkoutDir)
```
[the remark this answers is not in the document]
Comment (iemejia) [reply .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Spark.groovy:48]: thx, done

File: .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Spark.groovy [hunk 3]
```diff
@@ -40,10 +40,16 @@ PostcommitJobBuilder.postCommitJob('beam_PostCommit_XVR_Spark',
...
+              tasks(':runners:spark:3:job-server:validatesCrossLanguageRunner')
```
[the remark this answers is not in the document]
Comment (iemejia) [reply .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Spark.groovy:49]: good idea, done

File: runners/spark/spark_runner.gradle [hunk 4]
```diff
@@ -90,24 +168,27 @@ dependencies {
...
-  provided library.java.spark_core
-  provided library.java.spark_sql
-  provided library.java.spark_streaming
-  provided library.java.spark_network_common
+  provided "org.apache.spark:spark-core_$spark_scala_version:$spark_version"
+  provided "org.apache.spark:spark-network-common_$spark_scala_version:$spark_version"
+  provided "org.apache.spark:spark-sql_$spark_scala_version:$spark_version"
+  provided "org.apache.spark:spark-streaming_$spark_scala_version:$spark_version"
+  if (project.path == ":runners:spark:3") {
+    runtimeOnly "com.fasterxml.jackson.module:jackson-module-scala_2.12:2.10.2"
```
[the remark this answers is not in the document]
Comment (iemejia) [reply runners/spark/spark_runner.gradle:176]: :+1: missed that one.

File: runners/spark/spark_runner.gradle [hunk 5]
```diff
@@ -90,24 +168,27 @@ dependencies {
...
-  runtimeOnly library.java.jackson_module_scala
-  runtimeOnly "org.scala-lang:scala-library:2.11.8"
   testCompile project(":sdks:java:io:kafka")
   testCompile project(path: ":sdks:java:core", configuration: "shadowTest")
   // SparkStateInternalsTest extends abstract StateInternalsTest
   testCompile project(path: ":runners:core-java", configuration: "testRuntime")
   testCompile project(":sdks:java:harness")
   testCompile library.java.avro
-  testCompile library.java.kafka
+  testCompile "org.apache.kafka:kafka_$spark_scala_version:1.0.0"
```
[the remark this answers is not in the document]
Comment (iemejia) [reply runners/spark/spark_runner.gradle:191]: oups rebase issue, fixed, thx

File: runners/spark/spark_runner.gradle [hunk 6]
```diff
@@ -184,6 +265,7 @@ task validatesRunnerBatch(type: Test) {
     // SDF
     excludeCategories 'org.apache.beam.sdk.testing.UsesUnboundedSplittableParDo'
     // Portability
+    excludeCategories 'org.apache.beam.sdk.testing.UsesImpulse'
```
[the remark this answers is not in the document]
Comment (iemejia) [reply runners/spark/spark_runner.gradle:268]: again rebase issue, fixed.

File: .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Spark.groovy (hunk 2 above)
[the remark this answers is not in the document]
Comment (iemejia) [reply .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Spark.groovy:48]: fixed thx

File: release/src/main/scripts/build_release_candidate.sh [hunk 7]
```diff
@@ -287,7 +287,8 @@ if [[ $confirmation = "y" ]]; then
   done
 
   echo '-------------Generating and Pushing Spark job server image-------------'
-  ./gradlew ":runners:spark:job-server:container:dockerPush" -Pdocker-tag="${RELEASE}_rc${RC_NUM}"
+  ./gradlew ":runners:spark:2:job-server:container:dockerPush" -Pdocker-tag="${RELEASE}_rc${RC_NUM}"
+  ./gradlew ":runners:spark:3:job-server:container:dockerPush" -Pdocker-tag="${RELEASE}_rc${RC_NUM}"
```
[the remark this answers is not in the document]
Comment (iemejia) [reply release/src/main/scripts/build_release_candidate.sh:291]: thanks I missed the publish one

**Comment by iemejia** [comment]
Closing this to open the real PR in the upstream Apache Beam repo, thanks for this first preview review!

**Pull request closed without merging**